src/main/kotlin/io/libp2p/core/security/secio/SecioHandshake.kt

```diff
@@ -0,0 +1,155 @@
+package io.libp2p.core.security.secio
+
+import com.google.protobuf.Message
+import com.google.protobuf.Parser
+import io.libp2p.core.PeerId
+import io.libp2p.core.crypto.PrivKey
+import io.libp2p.core.crypto.PubKey
+import io.libp2p.core.crypto.StretchedKey
+import io.libp2p.core.crypto.keys.decodeEcdsaPublicKeyUncompressed
+import io.libp2p.core.crypto.keys.generateEcdsaKeyPair
+import io.libp2p.core.crypto.sha256
+import io.libp2p.core.crypto.stretchKeys
+import io.libp2p.core.crypto.unmarshalPublicKey
+import io.libp2p.core.types.compareTo
+import io.libp2p.core.types.toProtobuf
+import io.netty.buffer.ByteBuf
+import io.netty.buffer.Unpooled
+import kotlinx.coroutines.channels.ReceiveChannel
+import org.bouncycastle.crypto.digests.SHA256Digest
+import org.bouncycastle.crypto.digests.SHA512Digest
+import org.bouncycastle.crypto.macs.HMac
+import org.bouncycastle.crypto.params.KeyParameter
+import org.bouncycastle.jce.ECNamedCurveTable
+import org.bouncycastle.jce.provider.BouncyCastleProvider
+import spipe.pb.Spipe
+import java.security.SecureRandom
+import javax.crypto.Cipher
+import javax.crypto.spec.IvParameterSpec
+import javax.crypto.spec.SecretKeySpec
+
+data class SecioParams(
+    val nonce: ByteArray,
+    val permanentPubKey: PubKey,
+    val ephemeralPubKey: ByteArray,
+    val keys: StretchedKey,
+
+    val curveT: String,
+    val cipherT: String,
+    val hashT: String,
+
+    val mac: HMac,
+    val cipher: Cipher
+)
+
+/**
+ * Created by Anton Nashatyrev on 14.06.2019.
+ */
+class SecioHandshake(
+    private val inboundChannel: ReceiveChannel<ByteBuf>,
+    private val outboundChannel: suspend (ByteBuf) -> Unit,
+    val localKey: PrivKey,
+    val remotePeerId: PeerId?) {
+
+    val random: SecureRandom = SecureRandom.getInstanceStrong()
+    val nonceSize = 16
+    val ciphers = linkedSetOf("AES-128", "AES-256")
+    val hashes = linkedSetOf("SHA256", "SHA512")
+    val curves = linkedSetOf("P-256", "P-384", "P-521")
+
+    suspend fun doHandshake(): Pair<SecioParams, SecioParams> {
+        val nonce = ByteArray(nonceSize)
+        random.nextBytes(nonce)
+        val localPubKeyBytes = localKey.publicKey().bytes();
+
+        val proposeMsg = Spipe.Propose.newBuilder()
+            .setRand(nonce.toProtobuf())
+            .setPubkey(localPubKeyBytes.toProtobuf())
+            .setExchanges(curves.joinToString(","))
+            .setHashes(hashes.joinToString(","))
+            .setCiphers(ciphers.joinToString(","))
+            .build()
+
+        val remotePropose: Spipe.Propose = writeRead(proposeMsg)
```

Comment:
If we decide to use Netty only without coroutines, I would make this handler a finite state machine.